ansible/roles/atmo-gui-tweaks/tasks/desktop-background.yml

```diff
@@ -2,27 +2,31 @@
 - name: Copy desktop background to instance
   copy:
     src: "{{ background_img }}"
-    dest: /usr/share/backgrounds/cloud_background.jpg
+    dest: '{{ background_dest }}'
 
-- name: Change desktop background on CentOS
-  become: yes
-  become_user: "{{ ATMOUSERNAME }}"
-  command: >
-    gconftool-2 -s -t string /desktop/gnome/background/picture_filename "/usr/share/backgrounds/cloud_background.jpg"
-  when: ansible_distribution == "CentOS"
+- name: check if XFCE is desktop environment
+  stat:
+    path: '/usr/bin/xfce4-session'
+  register: xfce
 
-# Note: These tasks trick the os by replacing their defaults with our own instead of properly
-# setting the background because "Unable to autolaunch a dbus-daemon without a $DISPLAY for X11"
-- name: Change desktop background on Ubuntu
-  copy:
-    src: /usr/share/backgrounds/cloud_background.jpg
-    dest: /usr/share/backgrounds/default.jpg
-    remote_src: True
-  when: ansible_distribution == "Ubuntu" and ansible_distribution_major_version < "16"
+- name: desktop environment is XFCE
+  set_fact:
+    change_background_cmd: 'xfconf-query -c xfce4-desktop --property /backdrop/screen0/monitor0/workspace0/last-image --create --type string --set'
+  when: xfce.stat.exists
 
-- name: Change desktop background on Ubuntu 16+
-  copy:
-    src: /usr/share/backgrounds/cloud_background.jpg
-    dest: /usr/share/backgrounds/xfce/xfce-teal.jpg
-    remote_src: True
-  when: ansible_distribution == "Ubuntu" and ansible_distribution_major_version >= "16"
+- name: check if Gnome is desktop environment
+  stat:
+    path: '/usr/bin/gnome-session'
+  register: gnome
+
+- name: desktop environment is Gnome
+  set_fact:
+    change_background_cmd: 'gconftool-2 -s -t string /desktop/gnome/background/picture_filename'
+  when: gnome.stat.exists and not xfce.stat.exists
+
+# dbus-launch is used here to launch the dbus-daemon and appropriate dekstop session and gather the environment variables that it creates.
+# dbus-launch is generally run as part of user login, but that has not happened yet so it is necessary here.
+- name: Change desktop background
```

Comment:
Basically launching dbus is the wrong thing to do here-its definitely a hack. You need to explain not that we're launching dbus nor what dbus does, but that its a hack and **why** the hack is necessary. I realize that you may not have the holistic reason why its a hack.

We're creating a rogue dbus just so that gsettings will be satisfied. This is the wrong way to use gsettings, because an existing user session + windows would have its own dbus, and never see our updated setting. However, it's okay in this instance, because we nuke the users' desktop. The new desktop will get the updated settings from a file.